src/ast/passes/semantic_analyser.h

```diff
@@ -161,7 +162,14 @@ class SemanticAnalyser : public Visitor {
   // SemanticAnalyser.
   int func_arg_idx_ = -1;
 
-  std::map<Scope *, std::map<std::string, SizedType>> variable_val_;
+  struct variable_type {
+    SizedType type;
+    bool can_resize;
+    bool was_assigned;
+  };
+
+  std::map<Scope *, std::map<std::string, variable_type>> variable_type_;
+  std::map<Scope *, std::map<std::string, location>> declared_variables_;
```

Comment:
Sorry about being such a nitpicker but with the new `was_assigned` field, `variable_type` does not make sense as the type name.

What about this:
```suggestion
  struct variable {
    SizedType type;
    bool can_resize;
    bool was_assigned;
  };

  std::map<Scope *, std::map<std::string, variable>> variables_;
  std::map<Scope *, std::map<std::string, location>> variable_decls_;
```
or, maybe, we could even add `location` to `struct variable` and have just one map (not sure if that would make sense)?

Reply:
> Sorry about being such a nitpicker but with the new was_assigned field, variable_type does not make sense as the type name.

No worries! You're right, `variable` is a better name.

> or, maybe, we could even add location to struct variable and have just one map (not sure if that would make sense)?

We'd have to add in another field like `was_declared`. Also I think adding location to all the variables is unnecessary at this point.

Reply:
> We'd have to add in another field like `was_declared`. Also I think adding location to all the variables is unnecessary at this point.

Or make it `std::optional<location> decl_location` (with default value `std::nullopt`) but that's probably just an implementation detail.